test/unit/test_matrix.cpp

```diff
@@ -0,0 +1,365 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2019, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#include "dlaf/matrix.h"
+
+#include "gtest/gtest.h"
+#include "dlaf_test/util_matrix.h"
+#include "dlaf_test/util_types.h"
+
+using namespace dlaf;
+using namespace dlaf_test;
+using namespace matrix_test;
+using namespace testing;
+
+template <typename Type>
+class MatrixTest : public ::testing::Test {};
+
+TYPED_TEST_CASE(MatrixTest, MatrixElementTypes);
+
+std::vector<GlobalElementSize> sizes({{31, 17}, {29, 41}, {0, 1}, {3, 0}});
+std::vector<TileElementSize> block_sizes({{7, 11}, {13, 11}, {3, 3}});
+
+TYPED_TEST(MatrixTest, Constructor) {
+  using Type = TypeParam;
+  auto el = [](const GlobalElementIndex& index) {
+    SizeType i = index.row();
+    SizeType j = index.col();
+    return TypeUtilities<Type>::element(i + 0.001 * j, j - 0.01 * i);
+  };
+
+  for (const auto& size : sizes) {
+    for (const auto& block_size : block_sizes) {
+      Matrix<Type, Device::CPU> mat(size, block_size);
+
+      EXPECT_EQ(MatrixBase(size, block_size), mat);
+
+      set(mat, el);
+
+      CHECK_MATRIX_EQ(el, mat);
+    }
+  }
+}
+
+template <class T, Device device>
+void checkFromExisting(T* p, const matrix::LayoutInfo& layout, Matrix<T, device>& matrix) {
+  auto el = [](const GlobalElementIndex& index) {
+    SizeType i = index.row();
+    SizeType j = index.col();
+    return TypeUtilities<T>::element(i + 0.001 * j, j - 0.01 * i);
+  };
+  auto el2 = [](const GlobalElementIndex& index) {
+    SizeType i = index.row();
+    SizeType j = index.col();
+    return TypeUtilities<T>::element(-i + 0.001 * j, j + 0.01 * i);
+  };
+  auto ptr = [p, layout](const GlobalElementIndex& index) { return getPtr(p, layout, index); };
+  const auto& size = layout.size();
+
+  for (SizeType j = 0; j < size.cols(); ++j) {
+    for (SizeType i = 0; i < size.rows(); ++i) {
+      *ptr({i, j}) = el({i, j});
+    }
+  }
+
+  EXPECT_EQ(MatrixBase(layout), matrix);
+  CHECK_MATRIX_PTR(ptr, matrix);
+  CHECK_MATRIX_EQ(el, matrix);
+
+  set(matrix, el2);
+
+  for (SizeType j = 0; j < size.cols(); ++j) {
+    for (SizeType i = 0; i < size.rows(); ++i) {
+      if (el2({i, j}) != *ptr({i, j})) {
+        FAIL() << "Error at index (" << i << ", " << j << "): "
```

Comment:
In general, you can use every GoogleTest macro as a stream (see [doc](https://github.com/google/googletest/blob/master/googletest/docs/primer.md#assertions))

In this particular case, `ASSERT_EQ(...) << "additional message"`